### DISCUSSION TITLE
Getting a heightmap from a Voxelized mesh

### DISCUSSION THREAD

**Discussion by SofianeHamlaoui** [pyvista/pyvista#4670] [Q&A]
So, Now I'm receiving a 32x32 Voxelized mesh !

I want to get the elevation of all upper faces of the mesh, technically 1024 counting the floor as 0 elevation.

anyone tried or did this before ? 

Thank you in advance

**Comment by SofianeHamlaoui** [comment]
For the moment I succeeded getting the heightmap using the elevation values, but I only get the available faces ( not where the mesh is empty ) 
![image](https://github.com/pyvista/pyvista/assets/16967174/64fe89b6-b6fc-490e-b936-95afbb234ddb)

I'm getting 780 values instead of 1024. I tried adding a floor but this isn't working or I didn't do it properly.

Thank you in advance.

**Comment by SofianeHamlaoui** [comment]
When I try adding a floor I get this : 
![image](https://github.com/pyvista/pyvista/assets/16967174/bf68fe75-adff-4b65-bd6e-e2d27c2f7c3e)

**Comment by SofianeHamlaoui** [comment]
and the floor looks like this : 

```python
floor_width = 10.0
floor_length = 10.0
floor_height = 0.01


floor = pv.Cube(x_length=floor_width, y_length=floor_length, z_length=floor_height)
floor.translate([-floor_width / 2, -floor_length / 2, 0.0])

```
![image](https://github.com/pyvista/pyvista/assets/16967174/895b919a-26ac-49ce-a1ef-7d5da832f9d0)

**Comment by SofianeHamlaoui** [comment]
After adding the floor ( as I'm getting the data from a WFS Stream with a BBox, I used the Bbox to make a polygon and generate it with the mesh ) 

To get the elevation and get the height of each upper voxel ( the highest ) I have this : 

![image](https://github.com/pyvista/pyvista/assets/16967174/7fa79a78-9683-4c12-9941-58bcabd393c9)


Means I'll use the elevation of each cell_center of each upper face to use it as a reference for my heightmap ( doing some math ) 

To do so I made this code ( not the cleanest ) 

```python
np.set_printoptions(threshold=sys.maxsize)

mesh = pv.read("voxeled_withfloor.obj")

elv = mesh.elevation(preference="cell")
elv.compute_normals(cell_normals=True)
face_normals = elv.cell_normals
upward_face_indices = np.where(face_normals[:, 2] > 0)[0]
upper_face_mesh = elv.extract_cells(upward_face_indices)
cell_centers = upper_face_mesh.cell_centers()
elevation = cell_centers.elevation()
elevation_array = elevation["Elevation"]

num_arrays = len(elevation_array) // 32
string_array = elevation_array.astype(np.float64)

multidimensional_arrays = np.reshape(string_array[: num_arrays * 32], (num_arrays, 32))

pl = pv.Plotter()
centers = elv.cell_centers()

pl.add_mesh(elv, show_edges=True, line_width=1)
pl.add_mesh(cell_centers, color="r", point_size=5.0, render_points_as_spheres=True)
_ = pl.show_grid()

pl.show()
```

**Comment by SofianeHamlaoui** [comment]
I thought about another thing, is it possible to add a 32x32 floor and get the position of each cell and it's matched elevation ? just having ideas to solve and find a solution for this !

**Reply by tkoyama010** [reply]
@SofianeHamlaoui  Hi could you upload `voxeled_withfloor.obj` that we can try your code?

**Reply by SofianeHamlaoui** [reply]
So, things changed now ! 

I'm getting the data directly from the WFS flux, and doing and extrude using the received data that already contains the heights of each building. 

We kept having the voxels height to generate the heightmap so we're doing it in the Front now but if it's possible to have the height of each voxel it would be perfect... but looks almost impossible ! 

Here is an **voxeled_withfloor.obj** example you can use if you want so : https://gist.github.com/SofianeHamlaoui/3340bd9721d1deac528d8a54371fc33f

**Reply by SofianeHamlaoui** [reply]
any news ?

**Reply by tkoyama010** [reply]
Sorry for the late reply. I found it would be useful to have a filter method that, given a vector, would retrieve the faces with that orientation (`faces_with_direction`). We will consider if this is feasible.

**Reply by SofianeHamlaoui** [reply]
Well the issues I had with that, is to voxelize it properly, because yeah using Pyvista itsn’t working that well so I had to use binvox instead for that part of the process.

for the heightmap, still not working using python so I’m doing it in the front side in JS.

**Comment by SofianeHamlaoui** [comment]
@banesullivan 
Hi, I can send you some OBJ test files or just send you the access to the GeoServer I'm using for that directly to check why the is always an issue with the voxelization if you want so ! I really want to use Pyvista instead of other similar tools... 

**Case :** because now, what I'm doing is receiving the bbox, getting the data with height, extruding to have the right and exact building 3d mesh, joining all the building into 1 and voxelizing them. 

for the moment it's ( kind of working ) but still want to use Pyvista for the voxelization part. 

Thank you in advance.
